### ISSUE TITLE
MAAS reports status okay when service is down

### ISSUE THREAD

**Issue by BjoernT** [rcbops/rpc-openstack#1023]
##### What version (commit) do you encounter this issue on?

All

##### What type of issue is this (bug, feature, documentation, regression)?

Bug

##### What is the expected behavior?

When a service calls fails, some MAAS checks still report status okay which should be changed to not ok:

```
---- snip ----
# python /usr/lib/rackspace-monitoring-agent/plugins/memcached_status.py 172.31.239.241
status okay
---- sap ----
```

Additionally all metrics are missing, which can cause alarms to go off based on the missing response.
Isn't it better to report the metrics in any case, just zeroed out if the service fails to prevent other alarms from rising ?

##### What is actually happening?

See above

##### How can we reproduce this issue?

You get the idea

##### How can we verify your issue is addressed?

Described behavior stops

##### If this is a large issue, what can be done immediately to alleviate the issue while working on a longer-term solution?

Medium priority

##### What is your evaluation on the importance of this issue, and why?

- 

###### Remove branches (lines) you do not wish to target.
- [ ] Master
- [ ] liberty-12.2
- [ ] liberty-12.1
- [ ] kilo

**Comment by OchiengEd** [comment]
You can reproduce the issue by repeatedly running the memcached rackspace monitoring plugin on the cli until you fail to get the full response expected.

# python /usr/lib/rackspace-monitoring-agent/plugins/memcached_status.py <memcached_container_ip>

**Comment by JCallicoat** [comment]
I can replicate with memcached connections close to the configured max. I believe what is happening is memcached hits the max connections and a new connection is needed to send the 'stats' command, but that fails to connect and the plugin returns no metrics.

A quick and dirty test seems to show that is indeed the problem. In  `get_stats` on `memcache.Client` connection failure is silently ignored:

https://github.com/linsomniac/python-memcached/blob/master/memcache.py#L319-L320

I changed it locally to print an error:

```python
            if not s.connect():
                sys.stderr.write("OUCH: can't connect!\n")
                continue
```

Now I see that error when the plugin returns no metrics:

```bash
for i in {0..1000} ; do /usr/lib/rackspace-monitoring-agent/plugins/memcached_status.py 172.31.239.243 ; sleep 1 ; done
[...snip...]
status okay
metric memcache_api_local_status uint32 1
metric memcache_total_items uint64 168288 items
metric memcache_get_misses uint64 48486 cache_misses
metric memcache_get_hits uint64 2461857 cache_hits
metric memcache_total_connections uint64 1874640 connections
OUCH: can't connect!
status okay
metric memcache_api_local_status uint32 0
```

Reference issue #1025

**Comment by prometheanfire** [comment]
closed in favor of #1025
